### PR TITLE
PB-5946: Added logic for PVC cleanup validation

### PR DESCRIPTION
**Pull request by sabrarhussaini**
**What this PR does / why we need it**:
Added logic for PVC cleanup validation to make sure that the Px-backup namespace does not have any residual PVC's to ensure proper cleanup after runs.

**Which issue(s) this PR fixes** (optional)
Address' the issue: https://purestorage.atlassian.net/browse/PB-4636
Task:
https://purestorage.atlassian.net/browse/PB-5946
https://purestorage.atlassian.net/browse/PB-6548

**Special notes for your reviewer**:
Run details for BasicBackupCreation TC:
1. Positive case:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/5049

2. Negative case:
a. PVC is not present before the run, but present after run:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/5046
b. PVC present before run, but missing after run:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/5047
c. Same count, but mismatch in PVC names
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/5048

Pipeline with all tests:
**Vanilla S3:**
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/basic-system-test/job/px/job/s3/job/px-backup-system-test-vanilla/709/
**Vanilla NFS:**
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/basic-system-test/job/px/job/NFS/job/PX-Cloud-Vanilla-NFS/273/

### CODE REVIEW

File: tests/backup_helper.go [hunk 1]
```diff
@@ -8540,3 +8542,56 @@ func GetUpdatedKubeVirtVMSpecForBackup(scheduledAppContextsToBackup []*scheduler
...
+	if len(pvcListBeforeRun) != len(pvcListAfterRun) {
+		return fmt.Errorf("mismatch in pvc count")
+	}
+
+	// Check if the contents of the PVC lists are the same
+	for i := 0; i < len(pvcListBeforeRun); i++ {
+		if pvcListBeforeRun[i] != pvcListAfterRun[i] {
+			return fmt.Errorf("mismatch in pvc list before and after the run")
+		}
+	}
```
Comment (kphalgun-px) [inline tests/backup_helper.go:8595]: The order of PVCs might change, right?

File: tests/backup_helper.go [hunk 2]
```diff
@@ -8540,3 +8542,56 @@ func GetUpdatedKubeVirtVMSpecForBackup(scheduledAppContextsToBackup []*scheduler
 	}
 	return nil
 }
+
+// SetPVCListBeforeRun will set the value of global variable pvcListBeforeRun with the PVC names
+func SetPVCListBeforeRun() error {
```
Comment (kphalgun-px) [inline tests/backup_helper.go:8547]: How about creating a single function to return a list of PVCs and update the corresponding variable in its call sites? In this case BeforeSuite and AfterSuite.

```go
var _ = BeforeSuite(func() {
...

pvcListBeforeRun , err := GetPVCList()
log.FailOnError(err, "failed to list PVCs before run")

...
})
```

File: tests/backup_helper.go (hunk 1 above)
Comment (sabrarhussaini) [reply tests/backup_helper.go:8595]: Kindly correct me if I'm wrong, the order of PVC list might not change

File: tests/backup_helper.go (hunk 2 above)
Comment (sabrarhussaini) [reply tests/backup_helper.go]: @kphalgun-px I have used the global variables from backup helper and used the above approach as suggested. But I didn't re-declare the 'err' as it was already declared in the same block, and redeclaring was giving the error;

10:26:05  #28 23.68 ./backup_basic_test.go:278:7: err redeclared in this block
10:26:05  #28 23.68 	./backup_basic_test.go:252:8: other declaration of err
Please review.

Comment (kphalgun-px) [reply tests/backup_helper.go]: If it has already been declared, do we need to re-declare it? How about?

```go
var _ = BeforeSuite(func() {
...

pvcListBeforeRun, err = GetPVCList()
log.FailOnError(err, "failed to list PVCs before run")

...
})
```

Comment (sabrarhussaini) [reply tests/backup_helper.go]: @kphalgun-px I believe that's how the current code is.

**Review by mkoppal-px — CHANGES_REQUESTED** [review]
Please run the same on NFS pipeline

File: tests/backup_helper.go [hunk 3]
```diff
@@ -8540,3 +8542,39 @@ func GetUpdatedKubeVirtVMSpecForBackup(scheduledAppContextsToBackup []*scheduler
 	}
 	return nil
 }
+
+// GetPVCList will get the list of PVC's and returns the PVC names
+func GetPVCList() ([]string, error) {
```
Comment (mkoppal-px) [inline tests/backup_helper.go:8547]: Either make this function accept any namespace or name this function as `GetPVCListFromPxBackupNamespace()`
I would prefer the former.

File: tests/backup_helper.go [hunk 4]
```diff
@@ -8540,3 +8542,39 @@ func GetUpdatedKubeVirtVMSpecForBackup(scheduledAppContextsToBackup []*scheduler
...
+        pvcNameList = append(pvcNameList, pvc.Name)
+    }
+    return pvcNameList, nil
+}
+
+// ValidatePVCCleanup will compare the PVC list before and after the run
+func ValidatePVCCleanup(pvcListBefore, pvcListAfter []string) error {
+	// Check if the total number of the PVC's are the same
+	if len(pvcListBefore) != len(pvcListAfter) {
+		return fmt.Errorf("mismatch in pvc count")
```
Comment (mkoppal-px) [inline tests/backup_helper.go:8570]: If there is a mismatch I would like to see the difference between the original PVC list and the current one

Comment (sabrarhussaini) [reply tests/backup_helper.go]: @mkoppal-px I have done the requested changes, please review.

File: tests/backup_helper.go (hunk 3 above)
Comment (sabrarhussaini) [reply tests/backup_helper.go]: @mkoppal-px I have done the requested changes, please review.

File: tests/backup_helper.go [hunk 5]
```diff
@@ -8540,3 +8542,40 @@ func GetUpdatedKubeVirtVMSpecForBackup(scheduledAppContextsToBackup []*scheduler
...
+	if len(pvcListBefore) != len(pvcListAfter) {
+		fmt.Println("Expected PVCs:", strings.Join(pvcListBefore, ", "))
+		fmt.Println("Actual PVCs:", strings.Join(pvcListAfter, ", "))
+		return fmt.Errorf("mismatch in PVC count")
+	}
+	var mismatchedPVCs []string
+	for i := 0; i < len(pvcListBefore); i++ {
+		if pvcListBefore[i] != pvcListAfter[i] {
+			mismatchedPVCs = append(mismatchedPVCs, pvcListBefore[i])
+		}
```
Comment (mkoppal-px) [inline tests/backup_helper.go:8572]: This logic may not work always. What if the size of `pvcListAfter` is less than that of `pvcListBefore`
Also, it is better to print it in a way which say `pvc1 is present in pvcListBefore but not in pvcListAfter`

Look at this function and implement it accordingly - https://go.dev/play/p/V7n5b7d7HpN

Comment (sabrarhussaini) [reply tests/backup_helper.go:8572]: @mkoppal-px If the the size of `pvcListAfter `is less than that of `pvcListBefore`, the same logic would work;
We would get the error of mismatch in PVC count with both the PVC lists.
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/5042/console

I will modify the function to print only the mismatched PVC.

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: tests/backup_helper.go (hunk 5 above)
Comment (sabrarhussaini) [reply tests/backup_helper.go]: @mkoppal-px I have modified the logic accordingly and have updated the run details for all scenarios in the description. I didn't run for all testcases but instead just for BasicBackupCreation. Please advise.

**Comment by sabrarhussaini** [comment]
@mkoppal-px Run details after the changes:
**Vanilla S3:**
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/basic-system-test/job/px/job/s3/job/px-backup-system-test-vanilla/709/
**Vanilla NFS:**
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/basic-system-test/job/px/job/NFS/job/PX-Cloud-Vanilla-NFS/273/

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**